src/subdomains/generic/user/services/webhook/webhook-notification.service.ts

```diff
@@ -0,0 +1,104 @@
+import { Injectable } from '@nestjs/common';
+import { Cron, CronExpression } from '@nestjs/schedule';
+import { DfxLogger } from 'src/shared/services/dfx-logger';
+import { HttpService } from 'src/shared/services/http.service';
+import { DisabledProcess, Process } from 'src/shared/services/process.service';
+import { Lock } from 'src/shared/utils/lock';
+import { MailType } from 'src/subdomains/supporting/notification/enums';
+import { NotificationService } from 'src/subdomains/supporting/notification/services/notification.service';
+import { IsNull } from 'typeorm';
+import { UserData } from '../../models/user-data/user-data.entity';
+import { User } from '../../models/user/user.entity';
+import { KycWebhookData } from './dto/kyc-webhook.dto';
+import { PaymentWebhookData } from './dto/payment-webhook.dto';
+import { WebhookDto, WebhookType } from './dto/webhook.dto';
+import { WebhookRepository } from './webhook.repository';
+
+@Injectable()
+export class WebhookNotificationService {
+  private readonly logger = new DfxLogger(WebhookNotificationService);
+
+  constructor(
+    private readonly webhookRepo: WebhookRepository,
+    private readonly http: HttpService,
+    private readonly notificationService: NotificationService,
+  ) {}
+
+  @Cron(CronExpression.EVERY_5_MINUTES)
+  @Lock(1800)
+  async checkCryptoPayIn() {
+    if (DisabledProcess(Process.WEBHOOK)) return;
+    await this.sendOpenWebhooks();
+  }
+
+  async sendOpenWebhooks(): Promise<void> {
+    const entities = await this.webhookRepo.find({
+      where: { sentDate: IsNull() },
+      relations: { user: { wallet: true }, userData: { users: { wallet: true, userData: true } } },
+    });
+
+    entities.length > 0 && this.logger.verbose(`Sending ${entities.length} 'webhooks`);
+
+    for (const entity of entities) {
+      try {
+        entity.user
+          ? await this.triggerUserWebhook(entity.user, JSON.parse(entity.data), entity.type, entity.reason)
+          : await this.triggerUserDataWebhook(entity.userData, JSON.parse(entity.data), entity.type, entity.reason);
+
+        await this.webhookRepo.update(...entity.confirmSentDate());
+      } catch (e) {
+        this.logger.error(`Failed to send webhook ${entity.id}:`, e);
+      }
+    }
+  }
+
+  // --- HELPER METHODS --- //
+
+  private async triggerUserDataWebhook<T extends PaymentWebhookData | KycWebhookData>(
+    userData: UserData,
+    data: T,
+    type: WebhookType,
+    reason?: string,
+  ): Promise<void> {
+    for (const user of userData.users) {
+      await this.triggerUserWebhook(user, data, type, reason);
+    }
+  }
```

Comment:
I would create a webhook entity for every webhook to be sent (instead of looping here), so we can keep track of every single webhook. Then you could also remove the UserData field on the webhook.